### PR TITLE
Bootstrap Gizmos plugin

### PR DESCRIPTION
**Pull request by DiogoMendonc-a**

### CODE REVIEW

File: core/src/cubos/core/ecs/entity/hash.cpp
Comment (RiscadoA) [inline core/src/cubos/core/ecs/entity/hash.cpp:1]: Don't forget to move this to a different PR! If you want tips with the git shenanigans hmu on Discord

**Review by RiscadoA — CHANGES_REQUESTED** [review]

File: engine/samples/gizmos/main.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+{
+    Cubos cubos{argc, argv};
+    cubos.addPlugin(gizmosPlugin);
+    cubos.addPlugin(rendererPlugin);
+    cubos.startupSystem(mockCamera);
+    cubos.startupSystem(drawStartingLineSystem).tagged("sample.init").after("cubos.gizmos.init");
+    cubos.system(drawSystem);
+    cubos.run();
+}
+/// [Run]
```
Comment (RiscadoA) [inline engine/samples/gizmos/main.cpp:56]: Hide non relevant systems from the snippet, e.g. end it earlier and move the mock camera system

File: engine/samples/gizmos/main.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+static void mockCamera(Write<ActiveCameras> camera, Commands cmds)
+{
+    camera->entities[0] = cmds.create()
+                              .add(Camera{.fovY = 60.0F, .zNear = 0.1F, .zFar = 100.0F})
+                              .add(LocalToWorld{})
+                              .add(Position{{5.0F, 5.0F, -10.0F}})
+                              .add(Rotation{glm::quatLookAt(glm::vec3{0.0F, 0.0F, 1.0F}, glm::vec3{0.0F, 1.0F, 0.0F})})
+                              .entity();
+}
```
Comment (RiscadoA) [inline engine/samples/gizmos/main.cpp:21]: Why do we call this a mock camera?

**Review by RiscadoA — COMMENTED** [review]
Thanks for working on this!

I've left some comments, but its only a first pass, as I focused more on the public interface and didn't go very deep into the implementation yet.

File: engine/samples/gizmos/main.cpp (hunk 2 above)
Comment (DiogoMendonc-a) [reply engine/samples/gizmos/main.cpp:21]: Because i copy pasted it from somewhere and forgot to change the name :)

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+#include <cubos/engine/cubos.hpp>
+
+using cubos::core::gl::RenderDevice;
+using cubos::core::gl::ShaderPipeline;
+using cubos::engine::DeltaTime;
+
+namespace cubos::engine
+{
+    /// @brief Resource which draws gizmos.
```
[the remark this answers is not in the document]
Comment (DiogoMendonc-a) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:20]: pointing to the sample should on the plugin.hpp file, no?

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        void drawQueuedGizmos(DeltaTime deltaTime);
+
+    private:
+        ShaderPipeline mPipeline;    ///< Shader pipeline to be used when drawing gizmos
+        RenderDevice* mRenderDevice; ///< Active render device
+        glm::vec3 mColor;            ///< Currently set color
+        class GizmoBase;
+        class LineGizmo;
+        class BoxGizmo;
+        std::vector<std::shared_ptr<GizmoBase>> mGizmosVector; ///< Queued gizmos to be drawn.
```
[the remark this answers is not in the document]
Comment (DiogoMendonc-a) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:69]: The issue would be that we'd need a way to reference the `GizmosRenderer` resource from `Gizmos`, which would make us need an innit still.

File: engine/include/cubos/engine/gizmos/gizmos.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (RiscadoA) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:20]: Yeah I guess it is enough

File: engine/include/cubos/engine/gizmos/gizmos.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (RiscadoA) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:69]: You just need to move the logic out of the resource itself, instead of keeping the draw method in Gizmos for example.

[the remark this answers is not in the document]
Comment (DiogoMendonc-a) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:69]: Yeah, the draw method we can remove from Gizmos, but not the innit. The gizmos need to know the renderDevice and shaderPipeline when they are instantiated.

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+{
+    /// @brief Resource which draws gizmos.
+    ///
+    /// @ingroup gizmos-plugin
+    class Gizmos final
+    {
+    public:
+        /// @brief Initiates the Gizmos resource.
+        /// @param renderDevice Active render device.
+        void init(RenderDevice& renderDevice);
```
[the remark this answers is not in the document]
Comment (DiogoMendonc-a) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:28]: The issue would be "replace resource initialization functions with constructors"?

[the remark this answers is not in the document]
Comment (RiscadoA) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:28]: Its [this one](https://github.com/GameDevTecnico/cubos/issues/325).

[the remark this answers is not in the document]
Comment (DiogoMendonc-a) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:28]: I'm talking about the new one to be added

[the remark this answers is not in the document]
Comment (RiscadoA) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:28]: Maybe just add a comment there

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
+/// @file
+/// @brief Resource @ref cubos::engine::Gizmos.
+/// @ingroup gizmos-plugin
+
+#pragma once
+#include <vector>
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:6]:
```suggestion

#include <vector>
```

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+#include <glm/glm.hpp>
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:8]:
```suggestion
#include <glm/vec3.hpp>
```

At least in this header you don't need anything else. Might have to change the source file includes.

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        Primitive boxPrimitive;  ///< GL box information
+
+        /// @brief Sets up the GizmosRenderer to be used
+        /// @param renderDevice the current RenderDevice being used
+        void init(cubos::core::gl::RenderDevice* currentRenderDevice);
+
+    private:
+        void initLinePrimitive();
+        void initBoxPrimitive();
+    };
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:42]: Move this resource to a header in the src/ dir, as its never used outside.

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            bool decreaseLifespan(float delta)
+            {
+                mLifespan -= delta;
+                return mLifespan <= 0;
+            }
+
+            virtual ~GizmoBase() = default;
+        };
+
+        std::vector<std::shared_ptr<GizmoBase>> gizmosVector; ///< Queued gizmos to be drawn.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:109]: Why not define `GizmoBase` in a source file?

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    private:
+        glm::vec3 mColor; ///< Currently set color
+    };
+
+} // namespace cubos::engine
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:114]:
```suggestion
} // namespace cubos::engine

```

**Review by RiscadoA — CHANGES_REQUESTED** [review]

File: engine/include/cubos/engine/gizmos/gizmos.hpp (hunk 8 above)
Comment (DiogoMendonc-a) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:42]: Y'know, i considered doing that, but dismissed it as it was a resource. I forgot you could just include the local header in the plugin.cpp too :)

File: engine/include/cubos/engine/gizmos/gizmos.hpp (hunk 9 above)
Comment (RiscadoA) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:109]: Ouh, usually that wouldn't be necessary since pointers support forward declarations. Ig the STL doesn't support it :pensive:. If thats the case, at least make it private, and make this a friend of GizmosRenderer using `friend class GizmosRenderer;`

Comment (DiogoMendonc-a) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:109]: I would have to move the iteration code from a system to the GizmosRenderer class, is that preferable?

Comment (RiscadoA) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:109]: Whoops. Yeah lets leave it as is then.

Comment (RiscadoA) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:109]: Could you rename `GizmoBase` to `Gizmo` though? From outside we don't see the implementations anyway

Comment (DiogoMendonc-a) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:109]: Alrighty. Was trying to avoid confusion between Gizmo and Gizmos, but you're right, it's private

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+#include <cubos/core/gl/render_device.hpp>
+
+#include <cubos/engine/cubos.hpp>
+
+namespace cubos::engine
+{
+    class GizmosRenderer;
+
+    /// @brief Resource which queues for drawing gizmos, basic primitives useful for debugging and tools.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:19]:
```suggestion
    /// @brief Resource which queues commands for drawing gizmos, basic primitives useful for debugging and tools.
```
Might need to wrap the line too, otherwise the formatter will be mad

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        /// @param lifespan How long the line will be on screen for, in seconds. Defaults to 0, which means a single
+        /// frame.
+        void drawWireBox(const std::string& id, glm::vec3 corner, glm::vec3 oppositeCorner, float lifespan = 0.0F);
+
+        /// @brief Class that describes a type of gizmo
+        class Gizmo
+        {
+        protected:
+            const std::string& mId; ///< Gizmo identifier
+            glm::vec3 mColor;       ///< Color of the gizmo
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:58]: Protected section should be beneath the public section for consistency

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+            /// @brief decreases the time the gizmo has left before it is destroyed
+            /// @param delta the time in seconds since the last frame
+            bool decreaseLifespan(float delta)
+            {
+                mLifespan -= delta;
+                return mLifespan <= 0;
+            }
+
+            virtual ~Gizmo() = default;
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:81]: This should be before the constructor

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+                , mColor(color)
+                , mLifespan(lifespan)
+            {
+            }
+
+            /// @brief draws the gizmo to screen
+            /// @param renderer the GizmosRenderer in use
+            virtual void draw(GizmosRenderer renderer) = 0;
+
+            /// @brief decreases the time the gizmo has left before it is destroyed
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:73]:
```suggestion
            /// @brief Decreases the time the gizmo has left before it is destroyed.
```

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+            /// @param delta the time in seconds since the last frame
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:74]:
```suggestion
            /// @param delta Seconds since the last frame.
            /// @return Whether the gizmo should be destroyee.
```

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+        public:
+            Gizmo(const std::string& id, glm::vec3 color, float lifespan)
+                : mId(id)
+                , mColor(color)
+                , mLifespan(lifespan)
+            {
+            }
+
+            /// @brief draws the gizmo to screen
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:69]:
```suggestion
            /// @brief Draws the gizmo to screen.
```

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+            /// @param renderer the GizmosRenderer in use
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:70]:
```suggestion
            /// @param renderer Renderer.```

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+            virtual void draw(GizmosRenderer renderer) = 0;
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:71]: Shouldn't this receive the renderer by reference?

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+        /// @param from One of the ends of the line to be drawn.
+        /// @param to The other end of the line to be drawn.
+        /// @param lifespan How long the line will be on screen for, in seconds. Defaults to 0, which means a single
+        /// frame.
+        void drawLine(const std::string& id, glm::vec3 from, glm::vec3 to, float lifespan = 0.0F);
+
+        /// @brief Draws a filled box gizmo.
+        /// @param id Identifier of the gizmo.
+        /// @param corner One of the corners of the box to be drawn.
+        /// @param oppositeCorner The opposite corner of the box to be drawn.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:40]: Why not from/to for consistency?

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                mLifespan -= delta;
+                return mLifespan <= 0;
+            }
+
+            virtual ~Gizmo() = default;
+        };
+
+        std::vector<std::shared_ptr<Gizmo>> gizmosVector; ///< Queued gizmos to be drawn.
+    private:
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:85]:
```suggestion

    private:
```

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+            bool decreaseLifespan(float delta)
+            {
+                mLifespan -= delta;
+                return mLifespan <= 0;
+            }
+
+            virtual ~Gizmo() = default;
+        };
+
+        std::vector<std::shared_ptr<Gizmo>> gizmosVector; ///< Queued gizmos to be drawn.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:84]:
```suggestion
        std::vector<std::shared_ptr<Gizmo>> gizmos; ///< Queued gizmos to be drawn.
```

File: engine/include/cubos/engine/gizmos/plugin.hpp [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+#include <cubos/engine/gizmos/gizmos.hpp>
+
+namespace cubos::engine
+{
+    /// @defgroup gizmos-plugin Gizmos
+    /// @ingroup engine
+    /// @brief Used to draw gizmos helpful for debugging and tools.
+    ///
+    /// ## Resources
+    /// - @ref Gizmos - stores gizmos information.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/plugin.hpp:20]:
```suggestion
    /// - @ref Gizmos - used to queue gizmo draw commands.
```

File: engine/samples/gizmos/page.md [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+
+@snippet gizmos/main.cpp Adding plugin
+
+To draw a gizmo, all you need to do is to get a reference to the @ref cubos::engine::Gizmos resource, and then call the draw function on it for the gizmo you want to draw. Additionally, you can also call the @ref cubos::engine::Gizmos::color function to set the color for future gizmos. So, for example if you want to draw a line in a given system, all you need to do is the following:
+
+@snippet gizmos/main.cpp Start Up System
+
+This code will draw a red line from the top-left of the screen to the bottom right, and it will stay there for 10 seconds.
+
+In this other example, we draw a line, a box, and a wire box. Unlike the one in the previous example, this system is not a start-up system, so the draw functions get called every single frame. When this happensl, you should set the lifetime of a gizmo to 0, which means it will be drawn for a single frame only. This way we avoid drawing gizmos on top of identical ones that were already there, or in the case of moving gizmos, leaving a trail of old version behind them.
```
Comment (RiscadoA) [inline engine/samples/gizmos/page.md:17]: Instead of specifying a lifetime of 0, you should omit the argument, as it defaults to 0

File: engine/samples/gizmos/main.cpp [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+static void drawSystem(Write<Gizmos> gizmos)
+{
+    gizmos->color({0, 0, 1});
+    gizmos->drawLine("test line", {0, 0, 0}, {1, 1, 1}, 0);
+
+    gizmos->color({0, 1, 1});
+    gizmos->drawBox("test box", {0.4, 0.4, 0}, {0.6, 0.6, 1}, 0);
+
+    gizmos->color({1, 0, 1});
+    gizmos->drawWireBox("test box", {0.2, 0.7, 0}, {0.3, 0.1, 0}, 0);
```
Comment (RiscadoA) [inline engine/samples/gizmos/main.cpp:31]: Maybe explain here what having the same id means for the gizmos?

File: engine/src/cubos/engine/gizmos/gizmos_renderer.hpp
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/gizmos_renderer.hpp:1]: Rename file to `gizmos/renderer.hpp`

File: engine/src/cubos/engine/gizmos/gizmos.cpp [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+                              // back
+                              2, 7, 3, 3, 7, 4,
+                              // left
+                              3, 5, 0, 3, 4, 5,
+                              // bottom
+                              3, 1, 2, 3, 0, 1,
+                              // top
+                              5, 7, 6, 5, 4, 7};
+    boxPrimitive.ib = renderDevice->createIndexBuffer(sizeof(indices), indices, IndexFormat::UInt, Usage::Static);
+}
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/gizmos.cpp:103]: I would add a `gizmos/renderer.cpp` and define those there

File: engine/src/cubos/engine/gizmos/gizmos.cpp [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+    gizmosVector.push_back(
+        std::make_shared<LineGizmo>(id, corner, glm::vec3{corner[0], corner[1], oppositeCorner[2]}, mColor, lifespan));
+    gizmosVector.push_back(std::make_shared<LineGizmo>(id, glm::vec3{corner[0], oppositeCorner[1], corner[2]},
+                                                       glm::vec3{corner[0], oppositeCorner[1], oppositeCorner[2]},
+                                                       mColor, lifespan));
+    gizmosVector.push_back(std::make_shared<LineGizmo>(id, glm::vec3{oppositeCorner[0], corner[1], corner[2]},
+                                                       glm::vec3{oppositeCorner[0], corner[1], oppositeCorner[2]},
+                                                       mColor, lifespan));
+    gizmosVector.push_back(std::make_shared<LineGizmo>(id, glm::vec3{oppositeCorner[0], oppositeCorner[1], corner[2]},
+                                                       oppositeCorner, mColor, lifespan));
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/gizmos.cpp:156]: Why not call `drawLine` multiple times here?

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+#include <cubos/core/gl/render_device.hpp>
+
+#include <cubos/engine/cubos.hpp>
+
+namespace cubos::engine
+{
+    class GizmosRenderer;
+
+    /// @brief Resource which queues for drawing gizmos, basic primitives useful for debugging and tools.
+    ///
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:20]: I think you should explain here in the docs of the class the semantics of having two gizmo objects with the same Id. IIRC two gizmos with the same ID should behave as a single 'object', right?

File: engine/src/cubos/engine/gizmos/types/box_gizmo.hpp
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/box_gizmo.hpp:1]: Rename to box.hpp

File: engine/src/cubos/engine/gizmos/types/box_gizmo.hpp [hunk 28]
```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <glm/gtc/matrix_transform.hpp>
+
+#include <cubos/engine/gizmos/gizmos.hpp>
+
+/// @brief A gizmo that is a filled box
+class BoxGizmo : public cubos::engine::Gizmos::Gizmo
+{
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/types/box_gizmo.hpp:9]: Even though this is a private header, you should still put this inside the namespace I think1

**Review by RiscadoA — CHANGES_REQUESTED** [review]

File: engine/include/cubos/engine/gizmos/gizmos.hpp (hunk 19 above)
Comment (DiogoMendonc-a) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:40]: a box doesn't really go from a point to another like a line does

File: engine/samples/gizmos/main.cpp (hunk 24 above)
Comment (DiogoMendonc-a) [reply engine/samples/gizmos/main.cpp:31]: So far, nothing at all. I was thinking of only adding that when interaction was implemented

File: engine/include/cubos/engine/gizmos/gizmos.hpp (hunk 27 above)
Comment (DiogoMendonc-a) [reply engine/include/cubos/engine/gizmos/gizmos.hpp:20]: gizmos with the same ID will share interaction functionality. There isn't any interaction yet, so i'm not sure how to document that.

File: engine/include/cubos/engine/gizmos/plugin.hpp (hunk 22 above)
Comment (RiscadoA) [reply engine/include/cubos/engine/gizmos/plugin.hpp:20]: Nope, since its private to the implementation we should omit it

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+            /// @param renderer Renderer.
+            virtual void draw(GizmosRenderer& renderer) = 0;
+
+            /// @brief Decreases the time the gizmo has left before it is destroyed.
+            /// @param delta Seconds since the last frame.
+            bool decreaseLifespan(float delta)
+            {
+                mLifespan -= delta;
+                return mLifespan <= 0;
+            }
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:76]: Move this to the source file

File: engine/include/cubos/engine/gizmos/gizmos.hpp [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+        {
+        public:
+            virtual ~Gizmo() = default;
+
+            Gizmo(const std::string& id, glm::vec3 color, float lifespan)
+                : mId(id)
+                , mColor(color)
+                , mLifespan(lifespan)
+            {
+            }
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/gizmos.hpp:64]: Move to source file

File: engine/include/cubos/engine/gizmos/plugin.hpp [hunk 31]
```diff
@@ -0,0 +1,28 @@
...
+namespace cubos::engine
+{
+    /// @defgroup gizmos-plugin Gizmos
+    /// @ingroup engine
+    /// @brief Used to draw gizmos helpful for debugging and tools.
+    ///
+    /// ## Resources
+    /// - @ref Gizmos - used to queue gizmo draw commands.
+    /// -
+    ///
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/plugin.hpp:22]:
```suggestion
```

File: engine/include/cubos/engine/gizmos/plugin.hpp [hunk 32]
```diff
@@ -0,0 +1,26 @@
...
+#include <cubos/engine/gizmos/gizmos.hpp>
+
+namespace cubos::engine
+{
+    /// @defgroup gizmos-plugin Gizmos
+    /// @ingroup engine
+    /// @brief Used to draw gizmos helpful for debugging and tools.
+    ///
+    /// ## Resources
+    /// - @ref Gizmos - used to queue gizmo draw commands.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gizmos/plugin.hpp:20]: Missing documentation for the tags and plugin deps, checkout other plugins for examples

File: engine/src/cubos/engine/gizmos/plugin.cpp [hunk 33]
```diff
@@ -0,0 +1,60 @@
...
+#include <cubos/engine/gizmos/plugin.hpp>
+#include <cubos/engine/window/plugin.hpp>
+
+#include "renderer.hpp"
+
+using cubos::core::ecs::Read;
+using cubos::core::ecs::Write;
+
+using cubos::core::gl::RenderDevice;
+using cubos::core::gl::Stage;
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/plugin.cpp:15]: Unused

File: engine/src/cubos/engine/gizmos/plugin.cpp [hunk 34]
```diff
@@ -0,0 +1,60 @@
+#include <vector>
+
+#include <cubos/core/gl/render_device.hpp>
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/plugin.cpp:3]: Unnecessary

File: engine/src/cubos/engine/gizmos/plugin.cpp [hunk 35]
```diff
@@ -0,0 +1,60 @@
...
+{
+    gizmosRenderer->init(&(*window)->renderDevice());
+}
+
+void cubos::engine::gizmosPlugin(Cubos& cubos)
+{
+    cubos.addResource<Gizmos>();
+    cubos.addResource<GizmosRenderer>();
+
+    cubos.addPlugin(cubos::engine::windowPlugin);
```
Comment (RiscadoA) [inline engine/src/cubos/engine/gizmos/plugin.cpp:54]: Move plugin dependency above the add resource

**Review by RiscadoA — APPROVED** [review]
LGTM! Left some comments for minor changes, don't forget to push fixes for them before resolving the comments - if you have auto merge enabled it will just merge this when all conversations are resolved
